File: basemodule/src/main/java/zjm/cst/dhu/basemodule/MyApplication.java
package zjm.cst.dhu.basemodule;

import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.os.StrictMode;

import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;

import zjm.cst.dhu.basemodule.dagger2.component.ApplicationComponent;
import zjm.cst.dhu.basemodule.dagger2.component.DaggerApplicationComponent;
import zjm.cst.dhu.basemodule.dagger2.module.ApplicationModule;
import zjm.cst.dhu.basemodule.dagger2.module.NetworkModule;
import zjm.cst.dhu.library.utils.others.CrashHandler;

/**
 * Created by zjm on 2017/2/23.
 */

public class MyApplication extends Application {

    private static Context mContext;
    private ApplicationComponent mApplicationComponent;
    private RefWatcher mRefWatcher;

    @Override
    public void onCreate() {
        super.onCreate();
        mContext = getApplicationContext();
        setStrictMode();
        setCrashHandler();
        initStetho();
        initLeakCanary();
        setupInjector();
    }

    /**
     *
     */
    private void setStrictMode() {
        if (BuildConfig.DEBUG && Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder().detectAll().penaltyLog().build());
            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder().detectAll().penaltyLog().build());
        }
    }

    /**
     *
     */
    private void setCrashHandler() {
        CrashHandler crashHandler = CrashHandler.getInstance(this);
        Thread.setDefaultUncaughtExceptionHandler(crashHandler);
    }

    /**
     *
     */
    private void initStetho() {
        //Stetho.initializeWithDefaults(this);
    }

    /**
     *
     */
    private void initLeakCanary() {
        mRefWatcher = LeakCanary.install(this);
    }

    private void setupInjector() {
        mApplicationComponent = DaggerApplicationComponent.builder().
                applicationModule(new ApplicationModule(this))
                .networkModule(new NetworkModule(this)).build();
    }

    public ApplicationComponent getApplicationComponent() {
        return mApplicationComponent;
    }

    public static Context getContext() {
        return mContext;
    }

    public static RefWatcher getmRefWatcher(Context context){
        MyApplication myApplication=(MyApplication) context.getApplicationContext();
        return myApplication.mRefWatcher;
    }
}

File: basemodule/src/main/java/zjm/cst/dhu/basemodule/MyApplication.kt
package zjm.cst.dhu.basemodule

import android.app.Application
import android.content.Context
import android.os.Build
import android.os.StrictMode

import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher
import zjm.cst.dhu.basemodule.dagger2.component.ApplicationComponent

import zjm.cst.dhu.basemodule.dagger2.module.ApplicationModule
import zjm.cst.dhu.basemodule.dagger2.module.NetworkModule
import zjm.cst.dhu.basemodule.dagger2.component.DaggerApplicationComponent
import zjm.cst.dhu.library.utils.others.CrashHandler

/**
 * Created by zjm on 2017/2/23.
 */

class MyApplication : Application() {
    var applicationComponent: ApplicationComponent? = null
        private set
    private var mRefWatcher: RefWatcher? = null

    override fun onCreate() {
        super.onCreate()
        context = applicationContext
        setStrictMode()
        setCrashHandler()
        initStetho()
        initLeakCanary()
        setupInjector()
    }

    /**

     */
    private fun setStrictMode() {
        if (BuildConfig.DEBUG && Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder().detectAll().penaltyLog().build())
            StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder().detectAll().penaltyLog().build())
        }
    }

    /**

     */
    private fun setCrashHandler() {
        val crashHandler = CrashHandler.getInstance(this)
        Thread.setDefaultUncaughtExceptionHandler(crashHandler)
    }

    /**

     */
    private fun initStetho() {
        //Stetho.initializeWithDefaults(this);
    }

    /**

     */
    private fun initLeakCanary() {
        mRefWatcher = LeakCanary.install(this)
    }

    private fun setupInjector() {
        applicationComponent = DaggerApplicationComponent.builder().applicationModule(ApplicationModule(this))
                .networkModule(NetworkModule(this)).build()
    }

    companion object {

        var context: Context? = null
            private set

        fun getmRefWatcher(context: Context): RefWatcher? {
            val myApplication = context.applicationContext as MyApplication
            return myApplication.mRefWatcher
        }
    }
}
